keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -372,6 +430,32 @@ Based on the data collected from user clusters, vast majority (> 99%) of the k8s
 
 The current Endpoints API only includes a boolean state (Ready vs. NotReady) on individual endpoint. However, according to pod life cycle, there are more states (e.g. Graceful Termination, ContainerReary). In order to represent additional states other than Ready/NotReady,  a status structure is included for each endpoint. More condition types can be added in the future without compatibility disruptions. As more conditions are added, different consumer (e.g. different kube-proxy implementations) will have the option to evaluate the additional conditions. 
 
+- #### Why not use a CRD?
+
+**1. Protobuf is more efficient** Currently CRDs don't support protobuf. In our
+testing, a protobuf watch is approximately 5x faster than a JSON watch. We used
+pprof to profile 2 versions of kube-proxy using EndpointSlices and running on 2
+different nodes in a 150 node cluster as it scaled up to 15k endpoints. Over the
+15 minute window, kube-proxy with JSON used 17% more CPU time, with the
+difference in `StreamWatcher.receive` accounting for all of that. With protobuf
+enabled, that function took 1/5th the time of the JSON implementation.
+
+**2. Validation is too complex** Validation of addresses relies on addressType,
```

Comment:
@lavalamp @erictune @apelisse and all other CRD-ish folks.

IMO it should be possible to statically express more validation for CRDs - common types like "IP address" and "DNS subdomain" should get distinct validation and error messages.

Cross-field validation is more complex, but I'd like to make a case for simple forms of this to be statically expressed too.  e.g. `if("X", IPAddress, DNSDomain)`

Reply:
cc @DirectXMan12

Reply:
There's two angles for the "types like IP addresses" -- we can support more "format" values (used in OpenAPI to express complex built-in validation) in kube-apiserver, and some of this can be validated with regexes like:

```go
// this is not correct, but gets the point across
// +kubebuilder:validation:Pattern=`\d{3}\.\d{3}\.\d{3}\.\d{3}`
type IPAddress string

// this would be nice if we could do it
// +kubebuilder:validation:Format=ip-address
type IPAddress string
```

As for dependent cross field validation, IIRC some of that was explicitly rejected, but it might be worth re-evaluating.

cc @sttts as well

Reply:
@thockin 

would rather make it easier to write validation webhooks than to try to extend OpenAPI to do cross-field validation.

Reply:
cross-field valdiation is decently rare too, from what I've seen.  Common validation for IP addresses and DNS names would be useful though.